src/cart/crt_group.c

```diff
@@ -150,7 +150,86 @@ static d_hash_table_ops_t lookup_table_ops = {
 	.hop_rec_free		= li_op_rec_free,
 };
 
+struct crt_rank_mapping *
+crt_rm_link2ptr(d_list_t *rlink)
+{
+	D_ASSERT(rlink != NULL);
+	return container_of(rlink, struct crt_rank_mapping, rm_link);
+}
+
+
+static int
+rm_op_key_get(struct d_hash_table *hhtab, d_list_t *rlink, void **key_pp)
+{
+	struct crt_rank_mapping *rm = crt_rm_link2ptr(rlink);
+
+	*key_pp = (void *)&rm->rm_key;
+	return sizeof(rm->rm_key);
+}
+
+static uint32_t
+rm_op_key_hash(struct d_hash_table *hhtab, const void *key, unsigned int ksize)
+{
+	D_ASSERT(ksize == sizeof(d_rank_t));
+
+	return (unsigned int)(*(const uint32_t *)key %
+		(1U << CRT_LOOKUP_CACHE_BITS));
+}
+
+static bool
+rm_op_key_cmp(struct d_hash_table *hhtab, d_list_t *rlink,
+	      const void *key, unsigned int ksize)
+{
+	struct crt_rank_mapping *rm = crt_rm_link2ptr(rlink);
+
+	D_ASSERT(ksize == sizeof(d_rank_t));
+
+	return rm->rm_key == *(d_rank_t *)key;
+}
+
+static void
+rm_op_rec_addref(struct d_hash_table *hhtab, d_list_t *rlink)
+{
+	struct crt_rank_mapping *rm = crt_rm_link2ptr(rlink);
+
+	D_ASSERT(rm->rm_initialized);
+	D_MUTEX_LOCK(&rm->rm_mutex);
+	rm->rm_ref++;
+	D_MUTEX_UNLOCK(&rm->rm_mutex);
+}
+
+static bool
+rm_op_rec_decref(struct d_hash_table *hhtab, d_list_t *rlink)
+{
+	uint32_t		ref;
+	struct crt_rank_mapping	*rm = crt_rm_link2ptr(rlink);
+
+	D_ASSERT(rm->rm_initialized);
+	D_MUTEX_LOCK(&rm->rm_mutex);
+	rm->rm_ref--;
+	ref = rm->rm_ref;
```

Comment:
This could be merged with the previous line, although we should be migrating hash tables to use atomics anyway.

Reply:
Will change to  ref = --rm->rm_ref;
As to atomics, yes, i will add a ticket to change it in all hashtables in one go.